include/sys/zpl.h

```diff
@@ -32,12 +32,49 @@
 #include <linux/exportfs.h>
 #include <linux/falloc.h>
 #include <linux/file_compat.h>
+#include <linux/fs.h>
 #include <linux/parser.h>
 #include <linux/task_io_accounting_ops.h>
 #include <linux/vfs_compat.h>
 #include <linux/writeback.h>
 #include <linux/xattr_compat.h>
 
+#ifdef FS_PROJINHERIT_FL
+#define	ZFS_PROJINHERIT_FL	FS_PROJINHERIT_FL
+#else
+#define	ZFS_PROJINHERIT_FL	0x20000000
+#endif
+
+#ifdef FS_IOC_FSGETXATTR
+typedef struct fsxattr zfsxattr_t;
+
+#define	ZFS_IOC_FSGETXATTR	FS_IOC_FSGETXATTR
+#define	ZFS_IOC_FSSETXATTR	FS_IOC_FSSETXATTR
+#else
+struct zfsxattr {
+	__u32		fsx_xflags;	/* xflags field value (get/set) */
+	__u32		fsx_extsize;	/* extsize field value (get/set) */
+	__u32		fsx_nextents;	/* nextents field value (get)   */
+	__u32		fsx_projid;	/* project identifier (get/set) */
+	unsigned char	fsx_pad[12];
+};
+typedef struct zfsxattr zfsxattr_t;
+
+#define	ZFS_IOC_FSGETXATTR	_IOR('X', 31, zfsxattr_t)
+#define	ZFS_IOC_FSSETXATTR	_IOW('X', 32, zfsxattr_t)
+#endif
+
+static inline boolean_t
+zpl_is_valid_projid(uint64_t projid)
+{
+	zfsxattr_t fsx;
+
+	if ((sizeof (fsx.fsx_projid) == sizeof (uint32_t) && projid == -1U) ||
+	    (sizeof (fsx.fsx_projid) == sizeof (uint64_t) && projid == -1ULL))
```

Comment:
Just for case it is extended as 64-bits in the future. But if you think 32-bits is enough, I will remove the redundancy.